consensus/src/dag/adapter.rs

```diff
@@ -47,41 +48,56 @@ pub trait ProofNotifier: Send + Sync {
     async fn send_commit_proof(&self, ledger_info: LedgerInfoWithSignatures);
 }
 
+pub(crate) fn compute_initial_block_and_ledger_info(
+    ledger_info_from_storage: LedgerInfoWithSignatures,
+) -> (BlockInfo, LedgerInfoWithSignatures) {
+    // We start from the block that storage's latest ledger info, if storage has end-epoch
+    // LedgerInfo, we generate the virtual genesis block
+    if ledger_info_from_storage.ledger_info().ends_epoch() {
+        let genesis =
+            Block::make_genesis_block_from_ledger_info(ledger_info_from_storage.ledger_info());
+
+        let ledger_info = ledger_info_from_storage.ledger_info();
+        let genesis_qc = QuorumCert::certificate_for_genesis_from_ledger_info(
+            ledger_info_from_storage.ledger_info(),
+            genesis.id(),
+        );
+        let genesis_ledger_info = genesis_qc.ledger_info().clone();
+        (
+            genesis.gen_block_info(
+                ledger_info.transaction_accumulator_hash(),
+                ledger_info.version(),
+                ledger_info.next_epoch_state().cloned(),
+            ),
+            genesis_ledger_info,
+        )
+    } else {
+        (
+            ledger_info_from_storage.ledger_info().commit_info().clone(),
+            ledger_info_from_storage,
+        )
+    }
+}
+
 pub struct OrderedNotifierAdapter {
     executor_channel: UnboundedSender<OrderedBlocks>,
     storage: Arc<dyn DAGStorage>,
     parent_block_info: Arc<RwLock<BlockInfo>>,
+    highest_committed_anchor_round: Arc<RwLock<Round>>,
```

Comment:
i think we can just use atomicu64 instead?

Reply:
But you still need the Arc because it is shared with state_sync_trigger, right?

Reply:
Okay, we needed to cache the ledger info as well, so I introduced `LedgerInfoProvider` and a trait. New commit, lmk.